src/psyclone/tests/gocean1p0_opencl_test.py

```diff
@@ -779,6 +777,46 @@ def test_opencl_options_effects():
     assert "ierr = clFinish(cmd_queues(2))" not in generated_code
 
 
+@pytest.mark.parametrize("dist_mem", [True, False])
+@pytest.mark.usefixtures("kernel_outputdir")
+def test_multiple_command_queues(dist_mem):
+    ''' Check that '''
+    psy, _ = get_invoke("single_invoke_two_identical_kernels.f90", API, idx=0,
+                        dist_mem=dist_mem)
+    sched = psy.invokes.invoke_list[0].schedule
+
+    # Set the boundaries inside the kernel and each kernel in a different
+    # OpenCL queue (kernel1 in queue 2, kernel2 in queue3)
+    trans = GOMoveIterationBoundariesInsideKernelTrans()
+    for idx, kernel in enumerate(sched.coded_kernels()):
+        trans.apply(kernel)
+        kernel.set_opencl_options({'queue_number': idx+2})
+
+    # Apply OpenCL transformation
+    otrans = OCLTrans()
+    otrans.apply(sched)
+
+    generated_code = str(psy.gen)
+
+    kernelbarrier = '''
+      ierr = clFinish(cmd_queues(2))
+      ! Launch the kernel'''
+
+    haloexbarrier = '''
+      ierr = clFinish(cmd_queues(2))
+      CALL cu_fld%halo_exchange(depth=1)'''
+
+    if dist_mem:
```

Comment:
This example dispatches kernel 2 to queue 3 and should not have a barrier in this part of the code because it is the last thing here (no other kernel or halo exchange has a dependency waiting for it). However if OpenCL end_barrier is True (which it is by default), there will be a barrier before the end of the Invoke, which in this case is immediately after kernel 2.

I see this as a different functionality and is tested in "test_opencl_options_effects". In that test I move one kernel to queue 2 and check that the barrier is added at the end.

>Is there code somewhere that ensures we declare the cmd_queues array to be large enough? Please could you add a test for that if we don't already have one.

The cmd_queue array is dynamically allocated in FortCL and it uses the argument passed to `env_ocl_init()`. The test `test_psy_init` already test that if we put a kernel in queue number 5, the argument passed to initialise the OpenCL environment is a 5.

Reply:
Ooh, I'm impressed at us! :-)